[2.5단계 - 장기 리팩토링] 밍트(김명지) 미션 제출합니다.

Starlight258 opened the pull request:
안녕하세요 조앤!

먼저 머지 기간이 끝남에도 리팩토링한 코드에 대해 피드백을 남겨주신다고 하셔서 정말 감사드립니다!
시간에 쫓겨 미처 고려하지 못한 부분들을, 주말동안 반영해서 리팩토링할 수 있었어요! 😄
성장의 기회를 주셔서 감사합니다 🙇‍♀️🙇‍♀️

리팩토링한 부분은 아래와 같습니다.

## 1. 기물별 움직이는 전략을 enum에 모아두기
```java
public enum Piece {

    CANNON(
            new Movements(
                    new Movement(UP),
                    new Movement(RIGHT),
                    new Movement(LEFT),
                    new Movement(DOWN)
            ),
            MovementType.PALACE_CONSIDERATE,
            PathObstacleRule.MUST_JUMP_EXACTLY_ONE_OBSTACLE,
            7
    ),
    CHARIOT(
            new Movements(
                    new Movement(UP),
                    new Movement(DOWN),
                    new Movement(RIGHT),
                    new Movement(LEFT)
            ),
            MovementType.PALACE_CONSIDERATE,
            PathObstacleRule.CANNOT_PASS_OBSTACLES,
            13
    ),
```

- 기존에는 `Piece`라는 구체 클래스에서 `이동 전략`과 `장애물 전략`을 주입받도록 구현하였어요.
- 그런데 그런 방식의 문제점은, 기물별 움직이는 전략은 이미 고정이 되어있는데 전략을 선택하도록 하여 **잘못된 조합의 전략이 주입될 수 있었어요**.
- 따라서 피드백 주신 것처럼 기물 별 전략이 고정되도록 수정해보았습니다. 
- 이를 통해 기물들을 생성하는 `PieceFactory`에서 기물과 전략을 따로따로 주입하지 않아도 되고, 잘못된 전략이 주입되는 버그를 막을 수 있어서 명쾌해졌습니다!

## 2. 전략 객체 enum으로 수정
```java
public enum PathObstacleRule {

    CANNOT_PASS_OBSTACLES {
        @Override
        public void validatePathObstacles(final Path path, final Board board) {
            if (containsObstacles(path, board)) {
                throw new IllegalArgumentException("[ERROR] 경로에 기물이 존재하여 이동할 수 없습니다.");
            }
        }
    },
    MUST_JUMP_EXACTLY_ONE_OBSTACLE {
        private static final int OBSTACLE_JUMPING_THRESHOLD = 1;

        @Override
        public void validatePathObstacles(final Path path, final Board board) {
            final int count = countObstacles(path, board);
            if (count != OBSTACLE_JUMPING_THRESHOLD) {
                throw new IllegalArgumentException("[ERROR] 오직 하나의 기물만 뛰어넘을 수 있습니다.");
            }
            if (wouldJumpSameTypePiece(path, board)) {
                throw new IllegalArgumentException("[ERROR] 같은 종류의 기물을 뛰어넘거나 잡을 수 없습니다.");
            }
        }
    };
```
- 1번과 이어지는 내용인데, 기존에는 전략 인터페이스를 만들고, 각 구현체를 생성하는 방식이었어요.
- 그런데 그럴 경우에, 여러 기물들이 같은 전략을 공유함에도 매번 전략들이 생성되는 문제가 있었어요.
- 싱글톤으로 구현하기 위해 `enum`을 사용했습니다!

### 3. Dao 중복 코드 제거
- `Dao`(추상 클래스)
```java
   public abstract class BaseDao {
protected final <T> List<T> findAll(final String query, final StatementSetter setter,
                                        final ResultMapper<T> resultMapper) {
        final List<T> results = new ArrayList<>();
        try (final Connection connection = databaseConnectionProvider.getConnection();
             final PreparedStatement preparedStatement = connection.prepareStatement(query)) {

            if (setter != null) {
                setter.setParameters(preparedStatement, connection);
            }

            final ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                results.add(resultMapper.mapRow(resultSet, connection));
            }
            return results;

        } catch (final SQLException e) {
            throw new RuntimeException(e);
        }
    }
```
- 사용부분
```java
public class TurnDaoImpl extends BaseDao implements TurnDao {
    @Override
    public Turn selectCurrentTeam() {
        final var query = "SELECT current_team FROM turn LIMIT 1";
        return findOne(query, (resultSet, connection) -> {
            final String teamName = getTeamNameById(connection, resultSet.getInt("current_team"));
            return Turn.initialize(Team.from(teamName));
        });
    }
```
- 기존에는 기물에 대한 데이터를 관리하는 `PieceDao`과 차례(`Turn`)을 관리하는 `TurnDao` 사이의 중복 코드가 존재했었어요. 그리고 프로덕션 구현체와 테스트 구현체 사이에도 중복 코드가 많았어요. 또한 `Dao` 내부에서도 커넥션을 얻고 쿼리를 실행하는 코드 사이에 중복되는 코드가 굉장히 많이 존재했었어요. 

- 피드백 주신 내용 중에, `JdbcTemplate` 처럼 템플릿을 이용하여 중복 코드를 줄일 수 있다는 내용을 반영해서 리팩토링했습니다.
- `템플릿 메서드 패턴`을 이용해서 기본적인 뼈대를 정의하고, 하위 클래스에서 이를 사용하는 방식으로 하니 메서드가 어떤 역할을 하는지 명확해지고, 가독성이 좋아졌습니다.

그 외에도 잘 모르는 사람이 코드를 봐도 한눈에 파악할 수 있도록 리팩토링했습니다. 직관적인 클래스, 메서드명을 사용하고 복잡한 부분은 최대한 단순한 구조를 가지도록 수정했습니다.

반영한 부분에 대해 수정할 부분이 있거나, 개선할 부분이 있다면 피드백 부탁드립니다!!
다시 한번 감사드립니다 🙇‍♀️🙇‍♀️

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **신규 기능**
	- 보드 내 좌표와 이동 범위에 대한 제약 조건이 추가되어, 잘못된 입력이 줄어들고 게임 진행이 더욱 명확해졌습니다.
	- 이동 검증 로직이 개선되어, 유효하지 않은 명령 시 명확한 피드백을 제공합니다.
  
- **리팩터**
	- 게임의 턴 전환 및 데이터 처리 방식이 개선되어, 보다 안정적이고 유지보수하기 쉬운 구조로 재구성되었습니다.
	- 사용자 인터페이스 출력 메시지가 정리되어 게임 경험이 향상되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Starlight258 committed:
feat: DB 연결 오류시 예외 발생하도록 수정

Starlight258 committed:
feat: null 대신 예외 발생하도록 수정

Starlight258 committed:
feat: DB 제약조건 추가

- 좌표 위치에 대한 컬럼 범위를 설정한다.

Starlight258 committed:
refactor: 기물별 움직이는 전략은 고정되므로 enum 필드로 상수화

- Piece별 움직이는 전략들은 PieceType 별로 고정되므로, Piece에서 조합하기 보다 enum에 명시하는 것이 잘못된 전략이 주입되는 것을 막을 수 있다.

Starlight258 committed:
fix: Turn 초기화시 기존 Turn 데이터 제거하도록 수정

Starlight258 committed:
refactor: 사용하지 않은 코드 제거

Starlight258 committed:
refactor: 메서드 추출

Starlight258 committed:
refactor: 패키지 변경

Starlight258 committed:
refactor: enum 상수명 및 메서드명 수정

Starlight258 committed:
refactor: PathValidator 로직 Piece로 이동

- 단순히 PathValidator는 위임만 하고 있으므로, 데이터를 가지고 있는 Piece에서 수행하는 것이 더 명확하고 구조가 간단하다.

Starlight258 committed:
refactor: 메서드명 수정 및 사용하지 않는 메서드 제거

Starlight258 committed:
feat: 게임 종료 필드를 없애고 Optional.empty로 표현하도록 수정

- PieceMove가 게임 실행 로직과 기물 움직임 모두 가지고 있으므로, 기물의 움직임만 표현하도록 수정한다.

Starlight258 committed:
feat: PieceMove에서 중복 필드 제거

- isCaught는 caughtPiece != null 여부로 판단가능하다.

Starlight258 committed:
refactor: PieceMove에서 움직이는 부분과 잡힌 부분의 생성자 분리

Starlight258 committed:
refactor: Dao 중복 코드를 제거하기 위해 상위 추상 클래스인 BaseDao 생성

Starlight258 committed:
refactor: 템플릿 메서드 패턴을 이용하여 JDBC 코드 중복 제거

Starlight258 committed:
refactor: 프로덕션과 테스트 사이 DB 연결 중복 코드를 상속을 통해 제거

Starlight258 committed:
refactor: 상속시 오버라이드 목적이 아닌 메서드 final 추가

Starlight258 committed:
refactor: 메서드명 및 상수명 직관적으로 수정

Starlight258 committed:
refactor: 메서드명 및 파라미터명 수정

Starlight258 committed:
refactor: 메서드명 수정

Starlight258 committed:
fix: TestDatabaseProvider 생성되도록 수정

Starlight258 commented on src/test/java/janggi/infrastructure/TestDatabaseProvider.java:
```diff
@@ -0,0 +1,19 @@
+    private static final String CONFIG_PATH = "config.properties";
+
+    private static ProductionDatabaseProvider instance;
+
```
[fix: TestDatabaseProvider 생성되도록 수정](https://github.com/woowacourse/java-janggi/pull/173/commits/8da66a0f0a4389a739b20b61405bed4b94f9bc82)

Starlight258 replied:
[feat: 생성자 접근 제어자 수정](https://github.com/woowacourse/java-janggi/pull/173/commits/7e90720daad6f74e5eefe5b7a4f5fce85e4a3462)

Starlight258 committed:
refactor: 추상 클래스에서 구체 클래스에 의존하지 않도록 수정

Starlight258 commented on src/main/java/janggi/infrastructure/AbstractDatabaseConnectionProvider.java:
```diff
@@ -7,25 +7,18 @@
+    public final Connection getConnection() {
         final Properties properties = new Properties();
-        try (InputStream input = DBUtil.class.getClassLoader().getResourceAsStream(CONFIG_PROPERTIES)) {
+        try (InputStream input = ProductionDatabaseProvider.class.getClassLoader().getResourceAsStream(configPath)) {
```
[refactor: 추상 클래스에서 구체 클래스에 의존하지 않도록 수정](https://github.com/woowacourse/java-janggi/pull/173/commits/5f80a2fcbaba123221b320f8be2c317c2214b10a)

Starlight258 committed:
feat: 생성자 접근 제어자 수정

- 싱글톤 패턴의 경우 외부에서 직접적으로 생성자를 호출하는 것을 막아야하므로 생성자의 접근제어자를 private으로 막는다.

Starlight258 commented on src/main/java/janggi/dao/turn/TurnDaoImpl.java:
```diff
@@ -21,83 +21,37 @@ public static TurnDao getTurnDao() {
+        executeUpdate(query, (preparedStatement, connection) -> {
             preparedStatement.setInt(1, getTeamIdByName(connection, team.name()));
-            preparedStatement.executeUpdate();
-        } catch (final SQLException e) {
-            throw new RuntimeException(e);
-        }
+        });
```
아직 게임방이 하나라서 고려할 필요 X

seovalue commented:
이번에 남기는 Comment는 테스트보다는 비즈니스 코드에 집중해볼게요.
그리고 제 사견을 좀 많이 담아볼게요ㅎㅎ
한번 더 리팩터링하느라 고생 많으셨어요~! (따로 Merge나 Request Change 하진 않을게요!)

seovalue commented on src/docker/db/mysql/init/init.sql:
```diff
@@ -34,7 +34,9 @@ CREATE TABLE piece
     piecetype int references piecetype (piecetype_id),
-    PRIMARY KEY (piece_id)
+    PRIMARY KEY (piece_id),
+    CHECK (y BETWEEN 1 AND 10),
+    CHECK (x BETWEEN 1 AND 9)
```
CHECK는 저희만 그런지 모르겠지만 잘 사용하진 않더라고요
이게 비즈니스 코드에서 Validation Failed이 나면 따로 잡아서 관리해주기 편한데 (Ex. 생성자에서 막는다던지)
DB에서 에러가 나면 아마 다른 쿼리 오류랑 구분되지 않는 에러로 올 수도 있어요
그래서 DB에는 대부분 일반적인 Schema만 정의하고 꼭필요한 Validation은 코드단에서 주로 하는 편이에요

Starlight258 replied:
DB에서 발생하는 검증 에러는 쿼리 오류랑 구분이 되지 않아서 잘 하지 않는군요 !!

[feat: sql 검증 코드 제거](https://github.com/woowacourse/java-janggi/pull/173/commits/30e67782a6b4d3ec7d59cd7c34e83776f10b52cb)

seovalue commented on src/main/java/janggi/JanggiConsole.java:
```diff
@@ -43,12 +44,12 @@ public void start() {
             resultView.printOrder(currentTeam);
-            final PieceMove pieceMove = movePiece(players, currentTeam);
-            if (!pieceMove.wantRun()) {
+            final Optional<PieceMove> pieceMove = movePiece(players, currentTeam);
+            if (pieceMove.isEmpty()) {
```
이 부분 move에서 Optional을 반환한다는게 외부에서는 직관적으로 이해되진 않는 것 같아요
왜 empty가 나올 수 있지? 이런 생각?

Starlight258 replied:
`movePiece`에 대한 결과가 `Optional`인 것이 API 사용자 입장에서는 명확하지 않을 수 있겠네요!
`Optional` 대신 `MoveResult`라는 응답에 대한 상위 객체를 생성해서, 움직임 여부에 대한 결과를 반환하도록 수정했습니다!

[feat: Optional 대신 MoveResult라는 응답에 대한 상위 객체 생성](https://github.com/woowacourse/java-janggi/pull/173/commits/35ba09cd297650cfd9c17f8b0358092026858465)

seovalue replied:
좋네요 굿

seovalue commented on src/main/java/janggi/dao/BaseDao.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class BaseDao {
```
👍 와 멋진데요!!!
jpa나 jdbcTemplate도 이런 형태로 추상화를 해서 사용하고 있을거예요!
💯

seovalue commented on src/main/java/janggi/dao/BaseDao.java:
```diff
@@ -0,0 +1,129 @@
+        }
+    }
+
+    @FunctionalInterface
```
함수형 인터페이스는 어떤 역할을 해주나요?
어떤 경우에 함수형 인터페이스로 선언이 가능해요?

Starlight258 replied:
```java
 @FunctionalInterface
  protected interface StatementSetter {
      void setParameters(PreparedStatement preparedStatement, Connection connection) throws SQLException;
  }
```
여기서 함수형 인터페이스는 함수를 파라미터로 넘길 수 있게 해줍니다. (일급 객체로 취급합니다)

```java
  executeUpdate(query, (preparedStatement, connection) -> {
            preparedStatement.setInt(1, getTeamIdByName(connection, team.name()));
        });
```
즉 람다를 사용할 수 있게 해주는데, 필요한 이유는 메서드를 호출하는 사람이 `PreparedStatement`를 받아 파라미터를 설정할 수 있도록 해주기 위해서입니다. (템플릿 콜백 패턴, `ResultMapper`도 마찬가지입니다!) 이를 통해 익명 클래스를 사용하는 것보다 간결하게 코드를 작성할 수 있습니다.

함수형 인터페이스로 선언하기 위해서는 단 하나의 추상 메서드만 가져야합니다.

seovalue commented on src/main/java/janggi/domain/piece/Piece.java:
```diff
@@ -0,0 +1,200 @@
+import janggi.domain.players.Team;
+import java.util.Arrays;
+
+public enum Piece {
```
이번에 가장 큰 개선이 전반적으로 Enum으로 바꿔주신 점인 것 같아요 👍 

저는 Enum으로 관리하는게 편하면서도 그렇지 않은 경우도 있는 것 같아요.
enum은 사실 나열될 수 있는 값들을 관리하기 좋게 만들어주는 타입이라고 생각하긴 해서요ㅎㅎ

만약, 지금의 구현을 Enum이 아닌 각 PieceType 별 객체를 만들어주고 각각의 필드로 움직임과 전략을 선언해둔다면 지금과는 어떤 장/단점과 차이가 있을까요?
위 질문을 고민해보는 과정에서, `Cannon에는 ObstacleRule이 필요없어질 때`, 혹은 ObstacleRule을 두 개 이상 갖는 기물이 생겨날 때 등과 같이 다양한 확장 가능성을 염두에 두고 고민해보시면 좋을 것 같아요!!

Starlight258 replied:
좋은 질문 감사해요!

### enum으로 리팩토링한 이유
지금까지 리팩토링한 것을 살펴보면, step1에서는 `Piece`라는 추상 클래스를 상속받는 모든 기물 클래스를 생성해두었고, 
step2에서는 `Piece`를 구체 클래스로 두고, 전략을 주입받도록 구현했었어요.

결정적으로 클래스에서 `enum`으로 수정한 이유는, `상위 클래스`(추상 클래스나 합성을 하는 클래스)에서 `각 기물의 종류`(하위 클래스)를 알고 있는 부분이 문제라고 생각했어요.` isKing()`, `isJumpingPiece()`라는 메서드가  `pieceType`별로 달라지는데, `상위 클래스`에서 물어보는 것이 잘못된 추상화가 아닐까? 라는 생각이 들었어요. (마치 `instanceof` 처럼요) 따라서 `Piece`라는 클래스가 아닌 `PieceType`이라는 `enum`이 상위 클래스가 되는 것이 더 올바른 추상화라고 생각했고, 이렇게 리팩토링이 되었어요.

### 각 PieceType별 객체를 만든다면?
기물의 필드로 움직임과 전략을 선언해둔다면 좀 더 객체지향적인 코드가 될 수 있을 것 같아요. 현재는 포의 움직임이 변경될 때, `piece`라는 `enum` 파일을 수정해야하지만, 각 클래스를 만들면 `cannon` 클래스만 수정하면 되니까요.

제가 생각한 이 방법의 문제점은 각 `PieceType`별 객체를 만들어 전략을 생성자로 주입하도록 하면, 기물별 전략은 고정되어있는데 전략을 선택해서 주입받을 수 있게 해서 `적합하지 않은 전략`이 주입될 수 있는 것이었어요. (포가 다른 움직임 전략을 주입받을 수 있다면 버그가 발생하지 않을까?) 유연하게 전략을 열어둔다고 생각할 수도 있을 것 같은데, 저는 장기라는 게임의 확장성에서는 `"기물 종류가 같은데 다른 움직임을 할 수 있는 경우"`는 발생하지 않을 것으로 봤어요. 

각 객체 클래스에 전략을 상수로 설정하는 방식도 유연하다는 부분에서는 적합하지 않을 것 같고요.

또한 움직임과 전략만 가진 클래스들이 생성되는 것이 좋아보이지 않았던 것 같아요. 단지 각 움직임과 전략을 제공하는 역할만 하는 것이 좋은 객체인가? 라는 생각도 했어요.

### 움직임에 대한 변경 사항이 생긴다면?
말씀해주신 것처럼, `Cannon`에 `ObstacleRule`이 필요없어질 때나 두 개 이상 전략을 가질 경우 각 객체로 관리하는 것이 `enum`의 구조를 변경하는 것보다 훨씬 `유지보수 관점`에서 좋은 것 같아요. 그저 파일이 많아지는 것보다 `enum`으로 한 곳에서 관리하는게 낫지 않을까?라는 생각을 했는데, 움직임이 추가될 경우에는 각 객체로 나누는 것이 더 나을 것 같아요! 

하지만 기물들을 묶어주는 클래스가 필요할 거고, 상위 클래스에서 하위 클래스 종류를 알아야하는 상황이 무조건 존재(승패 결정을 위해 기물 종류가 왕인지 알아야함)하기 때문에 `PieceType`별 객체를 만들어 상속 또는 합성을 하는 방식으로 다시 돌아가서 리팩토링하지는 않을 것 같아요.

### 결론
지금은 `enum`을 사용하는 것이 추상화나 구조 측면에서 명확하다고 생각하지만, 만약 도메인의 규칙이 변경되어서 기물의 움직임 전략 여러개를 조합하게되거나, 기물별 상태를 관리해야할 때! 각 기물 클래스를 만들어 확장을 고려할 것 같아요! (`YAGNI`) 

---
덕분에 더 깊게 장기 게임에 대한 설계에 대해 더 깊게 고민하게 되었어요~ 감사해요 조앤!! 😄

seovalue commented on src/main/java/janggi/domain/piece/PiecesFactory.java:
```diff
@@ -0,0 +1,198 @@
+                new Position(SECOND_CHANGE_X, 9),
+                Piece.CHO_SOLDIER
+        );
+        return new Board(board);
```
요즘에 드는 생각은 이렇게 휴먼에러가 발생하기 좋은 부분에서는 테스트코드 말고 Board 생성자 안에서 validation을 걸어줘도 괜찮겠단 생각이 들어요.
Board 안에서 모든 piece가 다 주입됐는지, 등등?

Starlight258 replied:
Board 내에 검증 코드를 만들어서 생성자에서 확인하는 방식도 좋네요!!

Starlight258 commented:
- `Test`용 `DatabaseProvider` 사용할 필요 X
- `AbstractDatabaseConnectionProvider`는 왜 추상클래스로 수정

Starlight258 committed:
feat: sql 검증 코드 제거

- db에서 발생하는 예외는 쿼리 예외와 검증 예외를 구분하기 어렵기 때문에 잘 사용하지 않는다.
- 코드 단에서 이미 검증하므로 check 부분 코드를 제거한다.

Starlight258 committed:
feat: Optional 대신 MoveResult라는 응답에 대한 상위 객체 생성

- 움직임 여부에 대한 결과를 반환하여 명확하게 한다.

Starlight258 committed:
fix: sql 쿼리 수정

Starlight258 committed:
feat: AbstractDatabaseConnectionProvider에 abstract 키워드 추가

Starlight258 committed:
feat: 필요없는 TestDatabaseProvider 제거

Starlight258 commented:
@seovalue 조앤 !! 코드 리뷰해주셔서 감사합니다 🙇‍♀️
덕분에 설계, 구현 측면에서 부족한 부분을 피드백받고, tradeoff에 대해 고려할 수 있었어요!
그리고 level2를 진행하기 전 좀 더 깊게 생각하고, 장단점을 고려할 수 있는 사고를 배우게 되었어요! ☺️

많이 배웠습니다 !! 감사합니다 조앤! 🙇‍♀️
